lib/msal-common/src/response/ResponseHandler.ts

```diff
@@ -51,6 +54,50 @@ export class ResponseHandler {
             tenantId: idTokenObj.claims.tid,
         };
     }
+
+    private saveToken(originalTokenResponse: TokenResponse, authority: string, serverTokenResponse: ServerAuthorizationTokenResponse, clientInfo: ClientInfo): TokenResponse {
+        const tokenResponse = { ...originalTokenResponse };
+        // Set consented scopes in response
+        const requestScopes = ScopeSet.fromString(serverTokenResponse.scope, this.clientId, false);
+        tokenResponse.scopes = requestScopes.asArray();
+
+        const accessTokenCacheItems = this.cacheManager.getAllAccessTokens(this.clientId, authority);
+
+        for (let i = 0; i < accessTokenCacheItems.length; i++) {
+            const accessTokenCacheItem = accessTokenCacheItems[i];
+            if (accessTokenCacheItem.key.homeAccountIdentifier === tokenResponse.account.homeAccountIdentifier) {
+                const cachedScopes = ScopeSet.fromString(accessTokenCacheItem.key.scopes, this.clientId, false);
+
+                if(cachedScopes.intersectingScopeSets(tokenResponse.scopes)) {
+                    this.cacheStorage.removeItem(JSON.stringify(accessTokenCacheItem.key));
+                }
+            }
+        }
+
+        // Expiration calculation
+        const expiresIn = TimeUtils.parseExpiresIn(serverTokenResponse.expires_in);
+        const expiration = TimeUtils.now() + expiresIn;
+
+        // Save access token in cache
+        const accessTokenKey = new AccessTokenKey(authority, this.clientId, serverTokenResponse.scope, clientInfo.uid, clientInfo.utid, this.cryptoObj);
+        const accessTokenValue = new AccessTokenValue(serverTokenResponse.access_token, serverTokenResponse.id_token, serverTokenResponse.refresh_token, expiresIn.toString(), tokenResponse.account.homeAccountIdentifier);
+        this.cacheStorage.setItem(JSON.stringify(accessTokenKey), JSON.stringify(accessTokenValue));
+
+        // Save tokens in cache
+        tokenResponse.accessToken = serverTokenResponse.access_token;
+        tokenResponse.idToken = serverTokenResponse.id_token;
+        tokenResponse.refreshToken = serverTokenResponse.refresh_token;
+        tokenResponse.expiresOn = new Date(expiration * 1000);
+        return tokenResponse;
+    }
+
+    private getCachedAccount(accountKey: string): Account {
+        try {
+            return JSON.parse(this.cacheStorage.getItem(accountKey)) as Account;
+        } catch (e) {
+            return null;
```

Comment:
do we not want to do anything with this error here if it happens? i get that it's relatively easy to see by looking at this code that it's an error in json parsing, but still might be useful to do something with this besides `return null`

Reply:
I'm not sure what we want to do in most cases, but I think it's better to assume that a malformed account is null rather than throw an error and force the application to break. But if you have any alternate suggestions, I'm happy to consider them.

Reply:
I think it's worth to at least do some sort of `logger.error()`, no?

Reply:
I will add this, good call.

Reply:
Added in the pre alpha PR.